### ISSUE TITLE
Chess from_fen

### ISSUE THREAD

**Issue by Carbon225** [sotetsuk/pgx#971]
Hi, is the `pgx.chess._from_fen` function correctly implemented?

```python
state = pgx.chess._from_fen('rnbqkbnr/pppp1ppp/4p3/8/2P5/6P1/PP1PPP1P/RNBQKBNR b KQkq - 0 2')
print(state.current_player, state._turn)
# 0 1  # the current player is not loaded
```

If this function is not supposed to be used, that's fine but it could be really useful.
In my case, I want to train a model from human data (PGN -> FEN -> `pgx.State`).

**Comment by sotetsuk** [comment]
Hi, thank you for the comment.

### `pgx.chess._from_fen` is not exposed to users

`pgx.chess._from_fen` is developed only for unit-testing (see example usage in  [test_chess.py](https://github.com/sotetsuk/pgx/blob/main/tests/test_chess.py#L43))
So the function name starts with `_` to avoid users to touch `_from_fen`.

However, as your use case looks important, I keep this issue open and we may provide `from_fen` function to users in the future.
Also, I guess current `_from_fen` reasonably works (but not well tested).

### `state.current_player` has nothing to do with `state._turn`

As you may misunderstood the API, I also mention that `state.current_player` does NOT represent turn (white/black).
`current_player` is required to identify the player id over multiple parallel runs.
We'll provide the more detailed API description very soon. 

(In the case of `_from_fen` usage, it always set `current_player` as zero)

**Comment by Carbon225** [comment]
> Also, I guess current _from_fen reasonably works (but not well tested).

Exactly, I think with a little effort it could be added to the public chess API.

> state.current_player does NOT represent turn (white/black).

I don't agree, see this example:

```python
env = pgx.make("chess")
key = jax.random.PRNGKey(0)
key, subkey = jax.random.split(key)
state = env.init(subkey)
for _ in range(5):
    print(state.current_player)
    key, subkey = jax.random.split(key)
    action = act_randomly(subkey, state)
    state = env.step(state, action)
```

The output is 0 1 0 1 0

So `current_player` changes every turn and seems to point at the player who is moving next (white/black).

Knowing the current player is important to extract the reward:

```python
new_state = env.step(state, action)
action_reward = new_state.rewards[state.current_player]
```

To extract the reward for making a move we need the current player.
So the user could either track the current player externally or `env.step` could return only the reward for the player making the action.
In my implementation of AlphaZero I wrapped `env.init` and `env.step` in a [Gym-like API](https://gymnasium.farama.org/api/env/) which makes it easier to use later.

How should I get the reward if `current_player` is not supposed to be the turn?

**Comment by sotetsuk** [comment]
Thank you for your feedback :)


We are preparing a detailed document but here we show a quick illustrative example:

```py
import jax
import jax.numpy as jnp
import pgx


# Prepare agent A and B
#   Agent A: random player
#   Agent B: baseline player provided by Pgx
A = 0
B = 1

# load environment
env = pgx.make("go_9x9")
init_fn = jax.jit(jax.vmap(env.init))
step_fn = jax.jit(jax.vmap(env.step))

# Prepare random player
from pgx.experimental.utils import act_randomly
act_randomly = jax.jit(act_randomly)
# Prepare baseline model
# Note that it additionaly requires Haiku library
model_id = "go_9x9_v0"
model = pgx.make_baseline_model(model_id)

# Initialize the states
seed = 42
batch_size = 10
key = jax.random.PRNGKey(seed)
key, subkey = jax.random.split(key)
keys = jax.random.split(subkey, batch_size)
state = init_fn(keys)
print(f"Game index: {jnp.arange(batch_size)}")
print(f"Black player: {state.current_player}")  # Note: A = 0 and  B = 1
# in other words
print(f"A is black: {state.current_player == A}")
print(f"B is black: {state.current_player == B}")


R = state.rewards
while not (state.terminated | state.truncated).all():
    # Action of random player A
    key, subkey = jax.random.split(key)
    action_A = act_randomly(subkey, state)
    # greedy action of baseline model B
    logits, value = model(state.observation)
    action_B = logits.argmax(axis=-1)

    action = jnp.where(state.current_player == A, action_A, action_B)
    state = step_fn(state, action)
    R += state.rewards

print(f"Return of agent A = {R[:, A]}")
print(f"Return of agent B = {R[:, B]}")
```

So 

> The output is 0 1 0 1 0

This just happened (with prob=0.5). You can see some game index starts with `state.current_player == 1 (i.e., B)`

```
state = init_fn(keys)
print(f"Game index: {jnp.arange(batch_size)}")
print(f"Black player: {state.current_player}")  # Note: A = 0 and  B = 1
# Game id: [0 1 2 3 4 5 6 7 8 9]
# Black player: [1 1 0 1 0 0 1 1 1 1]
```

> So the user could either track the current player externally or env.step could return only the reward for the player making the action.

Return only a single reward (like PettingZoo API) was our API candidate but we finally decided to return rewards for all player to allow users to access like

```py
print(f"Return of agent A = {R[:, A]}")
print(f"Return of agent B = {R[:, B]}")
```

Note that we can avoid to explicitly deal with the first batch dimension like `[:, A]` by using `vmap` at last.

Please feel free to ask anything.

**Comment by Carbon225** [comment]
Thank you for the detailed response.

Sorry I'm still going on about this but I want to clear this up.

I finally understood what you meant.
The first player is randomly chosen at the beginning.
But this has nothing to do with black/white going first because black always goes first (in Go).
So current player is a mapping from "agents" to physical positions around the game table like black/white. It says which agent should make the decision but not whether black/white plays.
This was what confused me the most but I think I see now.
In your example the random player plays as black/white randomly but initially I thought it always played black.

There is one thing you might have misunderstood, in my example the "0 1 0 1 0" sequence is not random. Current player switches between 0 and 1 at each call to env.step, right?

Thank you for your help.

**Comment by Carbon225** [comment]
Correction: I don't play Go but it seems black does not always play first (with handicaps). That's irrelevant to my question though, in chess the starting colour is always the same.

**Comment by sotetsuk** [comment]
Thank you for your reply! I think your descriptions are correct.
Some minor comments:

> in my example the "0 1 0 1 0" sequence is not random.

I just meant that there are two possible sequences "0 1 0 1 0 1 ...." and "1 0 1 0 1 0 ..." because players play by turn in chess.
As you mentioned, it's not random (except the first `current_player`).

>  I don't play Go but it seems black does not always play first (with handicaps)

I think Go is supposed to start always with black player.

Thanks!

**Comment by Carbon225** [comment]
Thanks for all the help, my AlphaZero implementation now works.

I will leave this issue open if you decide to implement a `from_fen` loader for the state.

About Go, it seems to be my mistake:

From <https://en.wikipedia.org/wiki/Go_(game)#Rules>:
> Aside from the order of play (alternating moves, Black moves first or takes a handicap) ...

I noticed this and thought that maybe white can play first but it seems it's just a way to balance the game by adding extra points at the end.

**Comment by sotetsuk** [comment]
Close as not planned now.
